apps/xprof_nif/xprof_core_nif_tracer.c

```diff
@@ -0,0 +1,259 @@
+#include <erl_nif.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/time.h>
+
+/* NIF interface declarations */
+static int load(ErlNifEnv* env,
+                void** priv_data,
+                ERL_NIF_TERM load_info);
+
+static int upgrade(ErlNifEnv* env,
+                   void** priv_data,
+                   void** old_priv_data,
+                   ERL_NIF_TERM load_info);
+
+static void unload(ErlNifEnv* env,
+                   void* priv_data);
+
+/* The NIFs: */
+static ERL_NIF_TERM enabled(ErlNifEnv* env,
+                            int argc,
+                            const ERL_NIF_TERM argv[]);
+
+static ERL_NIF_TERM trace(ErlNifEnv* env,
+                          int argc,
+                          const ERL_NIF_TERM argv[]);
+
+static ERL_NIF_TERM trace_call(ErlNifEnv* env,
+                               int argc,
+                               const ERL_NIF_TERM argv[]);
+
+static ErlNifFunc nif_funcs[] = {
+        {"enabled", 3, enabled},
+        {"trace", 5, trace},
+        {"trace_call", 5, trace_call}
+};
+
+ERL_NIF_INIT(xprof_core_nif_tracer, nif_funcs, load, NULL, upgrade, unload)
+
+/* Local functions */
+
+static void read_mfa(ErlNifEnv * env,
+                     const ERL_NIF_TERM mfa_tuple,
+                     char * module,
+                     char * function,
+                     int  * arity)
+{
+    const ERL_NIF_TERM * tuple;
+    int                  tuple_arity;
+
+    enif_get_tuple(env, mfa_tuple, &tuple_arity, &tuple);
+    enif_get_atom(env, tuple[0], module,   256, ERL_NIF_LATIN1);
+    enif_get_atom(env, tuple[1], function, 256, ERL_NIF_LATIN1);
+    enif_get_int(env, tuple[2], arity);
+}
+
+static ERL_NIF_TERM create_mfa_tuple(ErlNifEnv * env,
+                             char * module,
+                             char * function,
+                             int    arity)
+{
+    return enif_make_tuple3(
+            env,
+            enif_make_atom(env, module),
+            enif_make_atom(env, function),
+            enif_make_int (env, arity)
+    );
+}
+
+static ERL_NIF_TERM mfa_to_bin(ErlNifEnv * env,
+                               const ERL_NIF_TERM mfa_tuple)
+{
+    const ERL_NIF_TERM * tuple;
+    int                  tuple_arity;
+
+    unsigned char module[256];
+    unsigned char function[256];
+    int           arity;
+
+    enif_get_tuple(env, mfa_tuple, &tuple_arity, &tuple);
+    enif_get_atom(env, tuple[0], module,   256, ERL_NIF_LATIN1);
+    enif_get_atom(env, tuple[1], function, 256, ERL_NIF_LATIN1);
+    enif_get_int(env, tuple[2], &arity);
+
+    unsigned char buffer[256];
+    sprintf(buffer, "%s:%s/%d", module, function, arity);

```

Comment:
maybe snprintf?. BTW `buffer` lenght is 256 and both `module` and `function` are limited to 256, so we might end up with overflow here.